File: code/generate_strategies/IR(pseudocode)/cpp/cpp/function_jsonschema_descend.cpp
#include <string>
#include <map>
#include <nlohmann/json.hpp>

class JSONValidator {
public:
    std::string validate_json(const std::string& data_str, const std::string& schema_str) {
        try {
            
            nlohmann::json data = nlohmann::json::parse(data_str);
            nlohmann::json schema = nlohmann::json::parse(schema_str);

            
            if (!data.is_object()) {
                return "Validation failed";
            }

            
            for (auto& [key, value] : data.items()) {
                auto props = schema.value("properties", nlohmann::json::object());
                auto it = props.find(key);
                if (it != props.end()) {
                    auto& field_schema = *it;
                    
                    if (field_schema.contains("type")) {
                        std::string type = field_schema["type"];
                        
                        bool type_valid = false;
                        if (type == "string") type_valid = value.is_string();
                        else if (type == "integer") type_valid = value.is_number_integer();
                        else if (type == "boolean") type_valid = value.is_boolean();
                        else if (type == "object") type_valid = value.is_object();
                        else if (type == "array") type_valid = value.is_array();
                        else if (type == "number") type_valid = value.is_number();
                        
                        if (!type_valid) {
                            return "Validation failed";
                        }
                    }
                    
                    if (value.is_object()) {
                        std::string nested_result = validate_json(value.dump(), field_schema.dump());
                        if (nested_result != "JSON is valid") {
                            return "Validation failed";
                        }
                    }
                } else if (!props.contains(key)) {
                    return "Validation failed";
                }
            }

            
            if (schema.contains("required")) {
                for (const auto& req : schema["required"]) {
                    if (!data.contains(req)) {
                        return "Validation failed";
                    }
                }
            }

            return "JSON is valid";
        }
        catch (const std::exception&) {
            return "Validation failed";
        }
    }
};

File: manual/data/python/src/function_jsonschema_descend.py
from jsonschema import exceptions

class JSONValidator:
    def validate_json(self, data_str: str, schema_str: str) -> str:
        try:
            import json
            data = json.loads(data_str)
            schema = json.loads(schema_str)

            if not isinstance(data, dict):
                return "Validation failed"

            type_map = {
                "string": str,
                "integer": int,
                "boolean": bool,
                "object": dict,
                "array": list,
                "number": (int, float)
            }

            for key, value in data.items():
                field_schema = schema.get("properties", {}).get(key)
                if field_schema:
                    if "type" in field_schema and not isinstance(value, type_map.get(field_schema["type"], object)):
                        return "Validation failed"
                    if isinstance(value, dict):
                        nested_result = self.validate_json(json.dumps(value), json.dumps(field_schema))
                        if nested_result != "JSON is valid":
                            return "Validation failed"
                elif key not in schema.get("properties", {}):
                    return "Validation failed"

            for req in schema.get("required", []):
                if req not in data:
                    return "Validation failed"

            return "JSON is valid"
        except Exception:
            return "Validation failed"
